### ISSUE TITLE
Support OSPF

### ISSUE THREAD

**Issue by danderson** [google/metallb#256]
OSPF would be an interesting addition to the protocol collection. In theory, speaking OSPF should be pretty simple, if we don't participate in the full link-state topology (i.e. we don't install routes, we are never on the path for any traffic except LB IPs...).

Step one is to read https://tools.ietf.org/html/rfc2328 and figure out if it's actually possible, and if so, how complex it is (do we have to support areas and other weird extensions?)

We will also have to support both OSPFv2 and OSPFv3 if we do this, for IPv6 support.

**Comment by foosinn** [comment]
I would love to see OSPF support. We have hardware that requires additional licenses for BGP, while OSPF is supported out of the box.

**Comment by mleklund** [comment]
A _LOT_ of cisco equipment can do OSPF out of the box but they want a plus license for BGP.  I suspect many of us might like this feature.

**Comment by danderson** [comment]
I understand that OSPF support would be handy, but I don't have the time to implement this myself right now. The issue is open so it's in the backlog of stuff to do, but I'm a solo developer and I can't prioritize this right now.

**Comment by mleklund** [comment]
Not a complaint!! You are doing great work, just wanted to chime in on the wish list.

**Comment by aojea** [comment]
I don't think that link-state routing protocols like OSPF are suitable for these scenarios, it's better to use distance vector protocols. If the problem is about licenses, it will be better and simpler to implement RIP

**Comment by vaizki** [comment]
The only reason I want OSPF support is because BGP does not officially (as in RFCs etc) support ECMP load-balancing where two equal cost routes can be added by BGP into the routing table. I know some (many?) vendors have proprietary ECMP modes for BGP but my network can't do it without swapping a lot of gear. So all the traffic to a certain LB IP always goes through a single k8s node right now even though there are multiple nodes announcing the IP.

RIP doesn't do ECMP either so doesn't help me at all. OSPF does ECMP very well but I understand fully that it is not a natural protocol for metallb.

**Comment by vaizki** [comment]
@aojea yes, exactly like that. This is also the correct behaviour according to OSPF, BGP and RIP specifications.
- OSPF will install both routes and do equal cost multiplath (ECMP) routing over them = loadbalance
- BGP will install one of the routes and stick with it, with the final tie-breaker being the IP address of the peer router
- RIP will either stay with one route or sometimes switch, depending on the implementation and how timers happen to coincide. But it will not use both routes at the same time

**Comment by aojea** [comment]
@vaizki thanks for the clarification, sometimes one is so used to the vendor's implementations that forget these things :)

OSPF seems complex to implement ... but RIP seems simple.

What about announcing the route using both protocols so the router can learn it via RIP and BGP, I think that configuring the same weight  you will have ECMP... or maybe I'm overcomplicating this.

**Comment by mleklund** [comment]
I was wondering if running bird or quagga as part of metallb for the routing portion might be a possibility.  It seems like it might make make long term maintenance easier as well, as someone else is maintaining the routing protocols.

**Comment by mleklund** [comment]
fair enough

**Comment by q3k** [comment]
I would also like for metallb to be able to talk OSPF to my ToR switches (Dell PC M6220 in an M1000e enclosure), as that switch only support OSPF and RIP. [1]

I'll be willing to help prototype and/or implement this in metallb. I was thinking of using https://github.com/bio-routing/bio-rd , but OSPF is not supported there yet - so that would be the first thing to sort out (not sure if rolling OSPF from scratch in metallb is very wise).

[1] - Interestingly, looking at firmware strings shows that BGP support is there somewhere, possibly just disabled behing a service command. There's no licensing scheme for that device, so it's not about that either.

**Comment by laoshancun** [comment]
LGTM.

**Comment by blackliner** [comment]
Since metallb uses FRR, shouldn't OSPF be supported out the box?

**Comment by xvilo** [comment]
I would like to add that this would simplify running homelab clusters backed with Unifi gear
